Indexing NDJSONs (#29)
meilisearch/specifications: text/0029-indexing-ndjson.md

- Title: Indexing NDJSON
- Start Date: 2021-04-12
- Specification PR: [PR-#29](https://github.com/meilisearch/specifications/pull/29)
- MeiliSearch Tracking-Issues: TBD

# Indexing NDJSON

## 1. Feature Description and Interaction

### I. Summary

> The initiation step of document indexing is to send some file matching a format to be parsed and tokenized in order to give search results to end-users. An [NDJSON](http://ndjson.org/) file is easier to use than a CSV file because it propose a convenient format for storing structured data.

Reviewer: ```suggestion
To index documents, the body of the add documents request has to match a specific format. That specific format is then parsed and tokenized inside MeiliSearch. After which, the documents added are in the pool of searchable and returnable documents.

An [NDJSON](http://ndjson.org/) file is easier to use than a CSV file because it propose a convenient format for storing structured data.
```

I had a hard time understanding the first part. Also the transition between the first part and the second part is confusing. What is the context of the comparison?

Author: I've updated that part.

### II. Motivation

Currently, the engine only accepts JSON format as a data source. We want to give users the possibility of another simple file format to use. Thus, give them more versatility at the data source choices for the indexation step.

> Writing performance is also considered as a motivation since JSON Lines file parsing is less CPU and memory intensive than parsing standard JSON because every new lines represent separate entries, making the NDJSON file streamable. Thus, more suited for indexing a consequent data set.

Reviewer: ```suggestion
Writing performance is also a motivation since JSON Lines file parsing is less CPU and memory-intensive than parsing standard JSON. When new lines represent separate entries it makes the NDJSON file streamable, thus, more suited for indexing a consequent data set.
```

Representing nested structures in a JSON object is easy and convenient.

### III. Additional Materials

TBD

### IV. Explanation

Newline-delimited JSON (`ndjson`), line-delimited JSON (`ldjson`), JSON lines (`jsonl`) are three terms expressing the same formats primarily intended for JSON streaming.

As of now, we will use `ndjson` in the next parts to refer to a file that represents JSON entries separated by a new line character.

- Each entries will represent a document for MeiliSearch.
- Each entries should be a valid JSON object.
- The file should be encoded in UTF-8.

#### Example of a valid NJSON

Given the NDJSON payload
'''
{"id":1, "label": "t-shirt", "price": 4.99, "colors": ["red", "green", "blue"]}
{"id":499, "label": "hoodie", "price": 19.99, "colors": ["purple"]}
'''
the search result should be displayed as
```json
{
  "hits": [
    {
      "id": 1,
      "label": "t-shirt",
      "price": 4.99,
      "colors": [
          "red",
          "green",
          "blue"
      ],
    },
    {
      "id": 499,
      "label": "hoodie",
      "price": 19.99,
      "colors": [
          "purple"
      ],
    }
  ],
  ...
}
```

#### API Endpoints

> > Each API endpoints mentioned above will now require a `application/x-ndjson` as `Content-Type` header to process NDJSON data.

Reviewer: ```suggestion
> Each API endpoints mentioned above will now require a `application/x-ndjson` as `Content-Type` header to be processed as NDJSON data.
```

#### Add or Replace Documents [📎](https://docs.meilisearch.com/reference/api/documents.html#add-or-replace-documents)

```curl
curl \
  -X POST 'http://localhost:7700/indexes/movies/documents' \
  -H 'Content-Type: application/x-ndjson' \
  --data '
    {"id":1, "label": "t-shirt", "price": 4.99, "colors": ["red", "green", "blue"]}\n
    {"id":499, "label": "hoodie", "price": 19.99, "colors": ["purple"]}
  '
```
> Response code: 202 Accepted

##### Error codes

> - Sending a different payload than the `Content-Type` header should return a `415 unsupported_media_type` error.
> - Too large payload according to the limit should return a `413 payload_too_large` error 
> - Wrong file encoding should return a `420 unprocessable_entity` error

> > - Invalid CSV data should return a `420 unprocessable_entity` error

Reviewer: Is this the right spec to talk about CSV? Or is it for context?

Author: Nice catch! I wanted to write NDJSON.

### Add or Update Documents [📎](https://docs.meilisearch.com/reference/api/documents.html#add-or-update-documents)

```curl
curl \
  -X PUT 'http://localhost:7700/indexes/movies/documents' \
  -H 'Content-Type: application/x-ndjson' \
  --data '
    {"id":1, "label": "t-shirt", "price": 4.99, "colors": ["red", "green", "blue"]}\n
    {"id":499, "label": "hoodie", "price": 19.99, "colors": ["purple"]}
  '
```
> Response code: 202 Accepted

##### Errors handling

> - Sending a different payload than the `Content-Type` header should return a `415 unsupported_media_type` error.
> - Too large payload according to the limit should return a `413 payload_too_large` error 
> - Wrong file encoding should return a `420 unprocessable_entity` error
> - Invalid NDJSON data should return a `420 unprocessable_entity` error

### V. Impact on documentation

> This feature should impact MeiliSearch users documentation by adding mention of `ndjson` capability inside Documents scope at [Add or replace documents](https://docs.meilisearch.com/reference/api/documents.html#add-or-replace-documents) and [Add or update documents](https://docs.meilisearch.com/reference/api/documents.html#add-or-update-documents).

Reviewer: ```suggestion
This feature should impact MeiliSearch users documentation by adding the possibility to use `ndjson` as a accepted format in the  Documents scope at [Add or replace documents](https://docs.meilisearch.com/reference/api/documents.html#add-or-replace-documents) and [Add or update documents](https://docs.meilisearch.com/reference/api/documents.html#add-or-update-documents).
```

> We should also not only mention JSON format in `unsupported_media_type` section on the [errors page](https://docs.meilisearch.com/errors/#unsupported_media_type) and add `ndjson` format.

Reviewer: Not sure I understand this sentence

Author: https://docs.meilisearch.com/errors/#unsupported_media_type The error part only mentions the use of the JSON format -> "Currently, MeiliSearch supports only JSON payloads."

We should also not only mention JSON format in `unsupported_media_type` section on the [errors page](https://docs.meilisearch.com/errors/#unsupported_media_type) and add `ndjson` format. The documentation says "Currently, MeiliSearch supports only JSON payloads."

Documentation should also guide the user in the correct way to properly format and send ndjson data. Adding a dedicated page for the purpose of formatting and sending ndjson data should be considered.

### VI. Impact on SDKs

This feature should impact MeiliSearch SDK's in the future by adding the possibility to send ndjson data to MeiliSearch on the previous explicited endpoints.

## 2. Technical Aspects

### I. Technical details

> ⚠ A missing Content-Type will be interpreted as `application/json` since it's the current behavior. Giving an `application/json` Content-Type leads to the same behavior.

Reviewer 2: Is it a technical detail? The users and the documentation should be aware of this. Technical details are for internal implementation from my POV.
This information is as important as any information in `Explanation`.

Author: Nice point @curquiza. I will update the specification according to this feedback!
